spec.html

```diff
@@ -32497,173 +32497,107 @@ <h1>String.prototype.replace ( _searchValue_, _replaceValue_ )</h1>
             1. Assert: Type(_replaceValue_) is String.
             1. Let _captures_ be a new empty List.
             1. Let _replacement_ be ! GetSubstitution(_searchString_, _string_, _position_, _captures_, *undefined*, _replaceValue_).
-          1. Return the string-concatenation of _preserved_, _replacement_, and the substring of _string_ from _position_ + _searchLength_.
+          1. Return the string-concatenation of _preceding_, _replacement_, and _following_.
         </emu-alg>
         <emu-note>
           <p>The `replace` function is intentionally generic; it does not require that its *this* value be a String object. Therefore, it can be transferred to other kinds of objects for use as a method.</p>
         </emu-note>
 
-        <emu-clause id="sec-getsubstitution" type="abstract operation">
+        <emu-clause id="sec-getsubstitution" type="abstract operation" oldids="table-replacement-text-symbol-substitutions,table-45">
           <h1>
             GetSubstitution (
               _matched_: unknown,
               _str_: unknown,
               _position_: a non-negative integer,
               _captures_: unknown,
               _namedCaptures_: unknown,
-              _replacement_: unknown,
+              _replacer_: unknown,
             )
           </h1>
           <dl class="header">
           </dl>
+          <p>For the purposes of this abstract operation, a <em>decimal digit</em> is a code unit in the range 0x0030 (DIGIT ZERO) to 0x0039 (DIGIT NINE) inclusive.</p>
           <emu-alg>
             1. Assert: Type(_matched_) is String.
             1. Let _matchLength_ be the number of code units in _matched_.
             1. Assert: Type(_str_) is String.
             1. Let _stringLength_ be the number of code units in _str_.
             1. Assert: _position_ &le; _stringLength_.
-            1. Assert: _captures_ is a possibly empty List of Strings.
-            1. Assert: Type(_replacement_) is String.
+            1. Assert: _captures_ is a possibly empty List, each of whose elements is a String or *undefined*.
+            1. Assert: Type(_replacer_) is String.
+            1. Let _preceding_ be the substring of _str_ from 0 to _position_.
             1. Let _tailPos_ be _position_ + _matchLength_.
-            1. Let _m_ be the number of elements in _captures_.
-            1. Let _result_ be the String value derived from _replacement_ by copying code unit elements from _replacement_ to _result_ while performing replacements as specified in <emu-xref href="#table-replacement-text-symbol-substitutions"></emu-xref>. These `$` replacements are done left-to-right, and, once such a replacement is performed, the new replacement text is not subject to further replacements.
-            1. Return _result_.
+            1. If _tailPos_ &ge; _stringLength_, then
+              1. Let _following_ be the empty String.
+            1. Else,
+              1. Let _following_ be the substring of _str_ from _tailPos_.
```

Comment:
Instead of an if-else, I'd probably write this as

```suggestion
              1. Let _following_ be the substring of _str_ from min(_tailPos_, _stringLength_).
```

Reply:
Well, except, if we *did* want to insert a Note, we might want to do it like so:
```
1. If _tailPos_ &gt; _stringLength_, then
  1. NOTE: This can only happen when ...
  1. Let _following_ be the empty String.
1. Else,
  1. Let _following_ be the substring of _str_ from _tailPos_.
```
(Note that this shifts the "=" case from the "then" to the "else" so that the Note doesn't apply, but that's valid, because it's an empty String either way.)

Reply:
If we go with the `min` phrasing, the note can read `NOTE: _tailPos_ can exceed _stringLength_ only when [...]`.

Reply:
I put in the `min` formulation, but I think it makes the step a bit hard to read (too much happening in one step). I think I'd prefer something like:
```
1. Let _tailPos_ be _position_ + _matchLength_.
1. If _tailPos_ &gt; _stringLength_, then
    1. NOTE: This can happen if ...
    1. Set _tailPos_ to _stringLength_.
1. Let _refReplacement_ be the substring of _str_ from _tailPos_.
```